arrow-libs/fx/arrow-collectors/src/commonMain/kotlin/arrow/collectors/Collectors.kt

```diff
@@ -0,0 +1,147 @@
+   *
+   * @param value Value returns as result of the collection.
+   */
+  public fun <R> constant(value: R): Collector<Any?, R> =
+    object : CollectorI<R, Any?, R> {
+      override val characteristics: Set<Characteristics> = Characteristics.IDENTITY_CONCURRENT_UNORDERED
+      override suspend fun supply(): R = value
+      override suspend fun accumulate(current: R, value: Any?) {}
+      override suspend fun finish(current: R): R = current
+    }
+
+  /**
+   * Counts the number of elements in the flow.
+   */
+  public val length: Collector<Any?, Int> =
+    object : CollectorI<AtomicInt, Any?, Int> {
+      override val characteristics: Set<Characteristics> = Characteristics.CONCURRENT_UNORDERED
+      override suspend fun supply(): AtomicInt = AtomicInt(0)
+      override suspend fun accumulate(current: AtomicInt, value: Any?) {
+        current.incrementAndGet()
+      }
+
+      override suspend fun finish(current: AtomicInt): Int = current.get()
+    }
+
+  /**
+   * Sum of all the values in the flow.
+   */
+  public val sum: Collector<Int, Int> =
+    intReducer({ 0 }, Int::plus)
+
+  public fun intReducer(
+    initial: () -> Int, combine: (Int, Int) -> Int,
+  ): Collector<Int, Int> =
+    object : CollectorI<AtomicInt, Int, Int> {
+      override val characteristics: Set<Characteristics> = emptySet()
+      override suspend fun supply(): AtomicInt = AtomicInt(initial())
+      override suspend fun accumulate(current: AtomicInt, value: Int) {
+        current.update { combine(it, value) }
+      }
+
+      override suspend fun finish(current: AtomicInt): Int = current.get()
+    }
+
+  public fun <M> reducer(
+    initial: () -> M, combine: (M, M) -> M, unordered: Boolean = false,
+  ): Collector<M, M> =
+    object : CollectorI<Atomic<M>, M, M> {
+      override val characteristics: Set<Characteristics> =
+        if (unordered) Characteristics.CONCURRENT_UNORDERED
+        else emptySet()
+
+      override suspend fun supply(): Atomic<M> = Atomic(initial())
+      override suspend fun accumulate(current: Atomic<M>, value: M) {
+        current.update { combine(it, value) }
+      }
+
+      override suspend fun finish(current: Atomic<M>): M = current.get()
+    }
+
+  /**
+   * Returns the "best" value from the value.
+   *
+   * @param selectNew Decides whether the new value is "better" than the previous best.
+   */
+  public fun <M> bestBy(
+    selectNew: (old: M, new: M) -> Boolean,
+  ): Collector<M, M?> =
+    object : CollectorI<Atomic<M?>, M, M?> {
+      override val characteristics: Set<Characteristics> = Characteristics.CONCURRENT_UNORDERED
+      override suspend fun supply(): Atomic<M?> = Atomic(null)
+      override suspend fun accumulate(current: Atomic<M?>, value: M) {
+        current.update { old ->
+          if (old == null) value
+          else if (selectNew(old, value)) value
+          else old
+        }
+      }
+
+      override suspend fun finish(current: Atomic<M?>): M? = current.get()
+    }
```

Comment:
I think the instance can be cached, and this suffers from nested nullability issue 🤔 

```suggestion
    ListCollector
    
   private object Empty
    
    private val ListCollector: CollectorI<Atomic<Any?>, M, M?> =
      object : CollectorI<Atomic<Any?>, M, M?> {
        override val characteristics: Set<Characteristics> = Characteristics.CONCURRENT_UNORDERED
        override suspend fun supply(): Atomic<Any?> = Atomic(Empty)
        override suspend fun accumulate(current: Atomic<Any?>, value: M) {
          current.update { old ->
            if (old == Empty) value
            else if (selectNew(old, value)) value
            else old
          }
        }

        override suspend fun finish(current: Atomic<Any?>): M? = current.get() as? M
      }
```